### PR TITLE
FTS: Ensure fresh results from a probe

### PR DESCRIPTION
**Pull request by l-wang**
This patch addresses two main scenarios:
1) Allowing multiple probes both internal and external to reuse the same
results when appropriate (ie: piggybacking on previous results). Multiple
requests should share the same results if they all request before the start of
a new fts loop, and after the results of the previous probe.

2) Ensuring fresh results from an external probe. When a request occurs during
a current probe in progress, this request should get fresh results rather
"piggybacking" or using the current results.

We use similar logic as the checkpointer code to detect whether a probe is in
progress with a probe start tick and probe end tick. To request a probe, we
send a signal requesting a fts results, then wait for a new loop to start, we
then wait again for that current loop to finish. This implementation uses a
busy wait loop. In the future, we can leverage the upstream conditaion variable
implementation which enables us to signal multiple fts notify processes.

Co-Authored-by: Alexandra Wang <lewang@pivotal.io>
Co-Authored-by: Jimmy Yih <jyih@pivotal.io>
Co-authored-by: Taylor Vesely <tvesely@pivotal.io>
Co-authored-by: Soumyadeep Chakraborty <sochakraborty@pivotal.io>
Co-authored-by: David Krieger <dkrieger@pivotal.io>

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [x] Document changes
- [x] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by ashwinstar** [comment]
Cool to see the README for FTS modified as part of this PR, super awesome folks! (didn't review the PR yet)

File: src/backend/cdb/cdbfts.c [hunk 1]
```diff
@@ -102,18 +102,51 @@ ftsUnlock(void)
 void
 FtsNotifyProber(void)
 {
-	Assert(Gp_role == GP_ROLE_DISPATCH);
-	uint8 probeTick = ftsProbeInfo->probeTick;
+	Assert(Gp_role == GP_ROLE_DISPATCH);;
```
Comment (dkrieger-pvtl) [inline src/backend/cdb/cdbfts.c]: `s/;;/;/`

File: src/include/cdb/cdbfts.h [hunk 2]
```diff
@@ -33,7 +33,9 @@
 typedef struct FtsProbeInfo
 {
 	volatile uint8		fts_statusVersion;
-	volatile uint8		probeTick;
+	volatile slock_t	fts_lck;
```
Comment (berlin-ab) [inline src/include/cdb/cdbfts.h]: Can we use the full word `lock` here? Is there a good reason to shorten it?

File: src/test/isolation2/isolation2_schedule [hunk 3]
```diff
@@ -165,6 +165,7 @@ test: segwalrep/dtm_recovery_on_standby
 test: segwalrep/commit_blocking_on_standby
 test: pg_basebackup
 test: pg_basebackup_with_tablespaces
+test: fts_manual_probeche
```
Comment (berlin-ab) [inline src/test/isolation2/isolation2_schedule]: this looks like it might be a typo

File: src/backend/cdb/cdbfts.c [hunk 4]
```diff
@@ -103,17 +103,50 @@ void
 FtsNotifyProber(void)
 {
 	Assert(Gp_role == GP_ROLE_DISPATCH);
-	uint8 probeTick = ftsProbeInfo->probeTick;
+	int			new_started,
```
Comment (berlin-ab) [inline src/backend/cdb/cdbfts.c]: Does postgres style have an opinion about declaring multiple variables in one statement?

File: src/test/isolation2/isolation2_schedule (hunk 3 above)
Comment (l-wang) [reply src/test/isolation2/isolation2_schedule]: Indeed. Good catch.

File: src/backend/cdb/cdbfts.c [hunk 5]
```diff
@@ -103,17 +103,50 @@ void
...
+	for (;;)
 	{
+		SpinLockAcquire(&ftsProbeInfo->fts_lck);
+		new_started = ftsProbeInfo->fts_probe_started;
+		SpinLockRelease(&ftsProbeInfo->fts_lck);
+
+		if (new_started != old_started)
+			break;
+
+		CHECK_FOR_INTERRUPTS();
```
Comment (berlin-ab) [inline src/backend/cdb/cdbfts.c]: Would there be any reason to check for interrupts if we found that the fts probe has started?

File: src/backend/cdb/cdbfts.c [hunk 6]
```diff
@@ -103,17 +103,50 @@ void
...
+
+		SpinLockAcquire(&ftsProbeInfo->fts_lck);
+		new_done = ftsProbeInfo->fts_probe_done;
+		SpinLockRelease(&ftsProbeInfo->fts_lck);
+
+		if (new_done - new_started >= 0)
+			break;
+
 		CHECK_FOR_INTERRUPTS();
+		pg_usleep(50000);
```
Comment (berlin-ab) [inline src/backend/cdb/cdbfts.c]: Is the sleep necessary?

File: src/backend/cdb/cdbfts.c (hunk 5 above)
Comment (l-wang) [reply src/backend/cdb/cdbfts.c]: I think we mostly want to check for interrupts before sleep, and we would always want to check for interrupt regardless the probe has started or not.

File: src/backend/cdb/cdbfts.c (hunk 6 above)
Comment (l-wang) [reply src/backend/cdb/cdbfts.c]: Given that we are implementing the busy waiting (poll and sleep) model instead of other synchronization mechanisms such as condition variable, the sleep is necessary. Otherwise this may starve the Fts process and the other probe request processes which also needs to acquire the lock to modify/read fts_probe_done and fts_probe_start.

File: src/backend/cdb/cdbfts.c (hunk 4 above)
Comment (l-wang) [reply src/backend/cdb/cdbfts.c]: The opinion is to either declare multiple variables of the same type in one line, or to declare them in multiple lines with the type keyword. I will change it the the latter. 

>When declaring variables in functions, declare them sorted by size (largest to smallest), then in alphabetical order; multiple ones per line are okay. 

>If a line overflows, reuse the type keyword.

-- https://man.openbsd.org/style

Postgres uses the bsd coding style https://wiki.postgresql.org/wiki/Developer_FAQ#What.27s_the_formatting_style_used_in_PostgreSQL_source_code.3F

File: src/include/cdb/cdbfts.h (hunk 2 above)
Comment (l-wang) [reply src/include/cdb/cdbfts.h]: I grep'd `slock_t.*_lck` and `slock_t.*lock` from the codebase, and `_lck` has more votes. Unless anyone as strong opinion on it we are going to stick with what we have now.

Comment (jchampio) [reply src/include/cdb/cdbfts.h]: By my count there are fifteen usages of `lock` or a variant thereof, but only five usages of `lck`?

Upstream Postgres doubles down, with dozens of `lock` variants. `mutex` is also in use.

Comment (l-wang) [reply src/include/cdb/cdbfts.h]: All right, earlier I didn't count `lock` without the `_` prefix. Will do `lock` then.

File: src/include/cdb/cdbfts.h [hunk 7]
```diff
@@ -33,8 +33,10 @@
 typedef struct FtsProbeInfo
 {
 	volatile uint8		fts_statusVersion;
-	volatile uint8		probeTick;
 	volatile uint8		fts_status[FTS_MAX_DBS];
+	volatile slock_t	fts_lock; /* protects the following counters */
+	volatile int64		fts_probe_started;
```
Comment (soumyadeep2007) [inline src/include/cdb/cdbfts.h]: Is there a reason why we are using an int64 for these counters? If we could tolerate an int32 we could use sig_atomic_t. That would remove the need for fts_lock perhaps.

(This is not terribly necessary but might simplify the code a little)

File: src/backend/cdb/cdbfts.c [hunk 8]
```diff
@@ -103,17 +103,50 @@ void
 FtsNotifyProber(void)
 {
 	Assert(Gp_role == GP_ROLE_DISPATCH);
-	uint8 probeTick = ftsProbeInfo->probeTick;
+	int			new_started;
+	int			old_started;
```
Comment (kalensk) [inline src/backend/cdb/cdbfts.c]: These receiver types should match the same type (or be larger) than the values they are recieving (ie: fts_probe_started and fts_probe_done).

File: src/backend/cdb/cdbfts.c [hunk 9]
```diff
@@ -103,17 +103,50 @@ void
...
 		pg_usleep(50000);
+	}
+
+	/*
+	 * We are waiting for fts_probe_done >= new_started, in a modulo sense.
+	 * Wait until current probe in progress is completed
+	 */
+	for (;;)
+	{
+		int			new_done;
```
Comment (kalensk) [inline src/backend/cdb/cdbfts.c]: This receiver type should match the same type (or be larger) than the values they are receiving (ie: fts_probe_done).

File: src/include/cdb/cdbfts.h (hunk 7 above)
Comment (jchampio) [reply src/include/cdb/cdbfts.h]: I think using `sig_atomic_t` would ensure atomic access but would still not provide the proper visibility guarantees. One of the tasks of a proper lock function is to tell the hardware not to play games with out-of-order execution, for example with memory fences et al. See [the top of `s_lock.h`](https://github.com/greenplum-db/gpdb/blob/f2e0802/src/include/storage/s_lock.h) for some discussion.

Comment (dkrieger-pvtl) [reply src/include/cdb/cdbfts.h]: I don't think `sig_atomic_t` will work.  See https://www.gnu.org/software/libc/manual/html_node/Atomic-Types.html.  The sense of `atomic` in `sig_atomic_t` there simply means "guaranteed to be read in a single instruction".  So you'd see the value at _some_ point in time, which could even be an uninitialized value(!).   You need atomicity, visibility and ordering to get the correct happens-before semantic needed for the behavior you expect.  Unfortunately, `atomic` is often used in interface names(like https://gcc.gnu.org/onlinedocs/gcc-4.3.5/gcc/Atomic-Builtins.html) that add visibility and ordering as well.  You'll note that most of our code actually uses `volatile sig_atomic_t` to get visibility and ordering.

In addition, this isn't performance critical as the ftsloop takes several seconds, and the manual probes happen rarely, mostly in testing AFAIK.  For testing, we'd like to atomically grab both counts, hence the spinlock.

As for `int64`, we knew that'd be large enough to prevent wraparound.  `uint32` or even `int32` should be as well but `int64` certainly will be...

File: src/backend/fts/README [hunk 10]
```diff
@@ -116,3 +116,49 @@ Fault Tolerance Service (FTS):
...
+mechanisms.
+
+
+              poll segments
+          +---------<--------+
+          |                  |  <-----+ request4
+          |     upper        |
+          |                  |
+          |                  ^
+      done|----------------- |start
```
[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply src/backend/fts/README]: Yeah, we'll clean up the diagram...it made sense to us ;).

File: src/test/isolation2/input/fts_manual_probe.source [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+!\retcode gpconfig -c gp_fts_probe_interval -v 3600;
+!\retcode gpstop -u;
+
+select gp_inject_fault_infinite('fts_before_probe', 'suspend', 1);
+
+-- start three probe requests and suspend before incrementing the start tick
+-- they should all share the same results
+1&: select gp_request_fts_probe_scan();
+2&: select gp_request_fts_probe_scan();
+3&: select gp_request_fts_probe_scan();
```
[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply src/test/isolation2/input/fts_manual_probe.source]: Nothing...which I think is the root of the problem in our tests.  We also need to clock the requests with faults as well.

[the remark this answers is not in the document]
Comment (l-wang) [reply src/test/isolation2/input/fts_manual_probe.source]: @jchampio What do you mean by waiting here? Do you mean these three requests might have not recorded the old `fts_probe_started` by the time the default session resumes the `fts_before_probe` fault in line 43?

[the remark this answers is not in the document]
Comment (jchampio) [reply src/test/isolation2/input/fts_manual_probe.source]: @l-wang Yes. Is that correct, or did I miss something that guarantees the happens-before relationship?

[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply src/test/isolation2/input/fts_manual_probe.source]: No, there 's nothing that maintains the happens-before for the clients relative to the loop fault points.  We need to diagram this out and make sure we provide synchronization so that all happens-before expectations are met by this test.

File: src/include/cdb/cdbfts.h (hunk 7 above)
Comment (asimrp) [reply src/include/cdb/cdbfts.h]: I vote for using uint8 and deal with wrap around rather than punting on it to use int64.  If we use `uint8` for both `fts_probe_started` and `fts_probe_done`, what do you think about the following refactor of `FtsNotifyProber()`?  (this is simply loud thinking, I have not even compiled that code)
```C
void
FtsNotifyProber(void)
{
    uint8 start, end;
    SpinLockAcquire(&ftsProbeInfo->fts_lock);
    start = ftsProbeInfo->fts_probe_started;
    end = ftsProbeInfo->fts_probe_done;
    SpinLockRelease(&ftsProbeInfo->fts_lock);

    if (start == end)
    {
        /*
         * FTS is not in the middle of a probe cycle.  New probe cycle
         * is about to start.  We only need to wait for that probe cycle to finish.
         */
        while (end == ftsProbeInfo->fts_probe_done)
        {
            CHECK_FOR_INTERRUPTS();
            pg_usleep(5000);
        }
    }
    else
    {
        Assert(start - end == 1);
        /*
         * FTS is in the middle of a probe cycle.  We must
         * wait until current current probe cycle is finished.
         */
        while (start == ftsProbeInfo->fts_probe_started)
        {
            CHECK_FOR_INTERRUPTS();
            pg_usleep(5000);
        }
        /*
         * We must also wait until the next probe cycle is finished.
         * Note: unsigned subtraction works properly with wraparound.
         * E.g. 255 + 1 = 0, 0 - 255 = 1, 1 - 255 = 2, and so on.
         */
        while (ftsProbeInfo->fts_probe_done - start < 2)
        {
            CHECK_FOR_INTERRUPTS();
            pg_usleep(5000);
        }
    }
}
```

Comment (asimrp) [reply src/include/cdb/cdbfts.h]: Minor comment on variable names: can we avoid the `fts_` prefix for new member variables we are adding?

File: src/backend/fts/README [hunk 12]
```diff
@@ -116,3 +116,49 @@ Fault Tolerance Service (FTS):
...
+          |                  |
+          v     lower        |
+          |                  |  <-----+ request3
+          |                  |  <-----| request2
+          +----------->------+  <-----+ request1
+                waitLatch
```
Comment (asimrp) [inline src/backend/fts/README]: SIGINT handler of FTS invokes `SetLatch()`.  That means, as soon as the first of the above request is received, FTS will terminate the wait and move on to start a new probe cycle.  Two or more requests cannot pile up as shown in the diagram.  Am I missing something?

File: src/backend/cdb/cdbfts.c (hunk 5 above)
Comment (asimrp) [reply src/backend/cdb/cdbfts.c]: +1, CHECK_FOR_INTERRUPTS() is good thing to do before sleep.  This is particularly useful when postmaster is performing shutdown and waiting for all children to exit.

File: src/backend/cdb/cdbfts.c (hunk 9 above)
Comment (asimrp) [reply src/backend/cdb/cdbfts.c]: +1

File: src/include/cdb/cdbfts.h (hunk 7 above)
Comment (jchampio) [reply src/include/cdb/cdbfts.h]: @asimrp You need the locks to guarantee proper visibility and ordering; see my and David's comments above. `volatile` is not enough, nor is `sig_atomic_t`.

File: src/backend/cdb/cdbfts.c [hunk 13]
```diff
@@ -103,17 +103,50 @@ void
 FtsNotifyProber(void)
 {
 	Assert(Gp_role == GP_ROLE_DISPATCH);
-	uint8 probeTick = ftsProbeInfo->probeTick;
+	int			new_started;
+	int			old_started;
+
+	SpinLockAcquire(&ftsProbeInfo->fts_lock);
+	old_started = ftsProbeInfo->fts_probe_started;
```
[the remark this answers is not in the document]
Comment (jchampio) [reply src/backend/cdb/cdbfts.c]: Atomicity is only one piece of the puzzle for proper concurrency; see [this conversation above](https://github.com/greenplum-db/gpdb/pull/8228#discussion_r307841926).

[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply src/backend/cdb/cdbfts.c]: Yes, as long as the read/write variable is aligned(i.e. if you pack a structure to have an int32 on a non-multiple of 4, the compiler should generate two loads/stores).  In fact, on a 64-bit architecture, 64-bit aligned read/write will be atomic.  But(see https://github.com/greenplum-db/gpdb/pull/8228#discussion_r307945879), you still need `volatile` for visibility and ordering for proper synchronization(i.e. a well-defined happens-before relationship).

But I think you are asking why we even need a lock on a `volatile` int32.   I think you are correct in general, but in our case...as long as there is only a single thread writing to the variables(in our case, FtsLoop()), and we do not need to deal with atomic writes/reads from sets of variables, we should be fine.   We do this for our testing, but we use fault points to insert "barriers" so I don't think we'd see it there.

_However_, not using a lock relies on subtle reasoning here, since we are really coordinating `FtsLoop` and `FtsNotifyProber`.  Our code in `FtsNotifyProber`  requires that`FtsNotifyProber::SendPostmasterSignal()` is AFTER we grab the FtsLoop start count, and that being guaranteed requires some sequence point to respect the C `as-if` rule for the notify thread.  Now, `SendPostMasterSignal::PMSignalState` _is_ volatile, but do we want to depend on that?

So, in short, I think it would be correct to not use a spinlock to protect the `start` and `done` counts, but it is fragile code.  And this is not performance critical...we expect very low contention on this lock as manual probes are rare.

Thoughts?

File: src/backend/fts/README (hunk 12 above)
Comment (dkrieger-pvtl) [reply src/backend/fts/README]: We will improve the diagram.  What we really mean is that these three request happen quickly in succession, so fast that they are all requested BEFORE the FtsLoop continues past wakeup(actually, before WakeUp and ResetLatch).  Also, note a lot of the difficulty in the testing from this PR is from the meaning of the phrase `as soon as the first of the above request is received, FTS will terminate the wait and move on to start a new probe cycle.`.  The notifier sends a signal to the postmaster, which then sends a signal to the fts main, whose signal handler releases the latch...it's asynchronous.

We should really have drawn it this way:
```
          |                  |   <------+ one net request
          |                  |  
          +----------->------+  <-----+ request1, request2, requeset 3
                waitLatch
```

**Comment by dkrieger-pvtl** [comment]
You suggest using `uint8_t` for the start/done variables in https://github.com/greenplum-db/gpdb/pull/8228#discussion_r308698348.  Note it is critical to not having a lost loop to `SendPostMasterSignal` after grabbing start/end and before checking the loop.  

Otherwise, I think the implementation is correct, but what do we gain by using it? We can also deal with wraparound with an `int32_t` and I find the reasoning using that simpler.

But do we need to deal with wrap-around?  A `uint32` is 4 billion loops, which is 100 years at a 1Hz fts probe rate.  Seems enough to me...

File: src/test/isolation2/input/fts_manual_probe.source (hunk 11 above)
[the remark this answers is not in the document]
Comment (kalensk) [reply src/test/isolation2/input/fts_manual_probe.source]: Should be fixed in our recent commit that rewrites the test to be more reliable, consistent, and faster.

File: src/include/cdb/cdbfts.h (hunk 7 above)
Comment (dkrieger-pvtl) [reply src/include/cdb/cdbfts.h]: See the currently implementation for our response to these.  In particular, we are using a `uint32` for the counts, but do not check for overflow since that would require over 4 billion calls to the manual probe.  But we could add it, but it does not seem necessary.

File: src/backend/fts/README (hunk 10 above)
[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply src/backend/fts/README]: See update.

File: src/backend/cdb/cdbfts.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply src/backend/cdb/cdbfts.c]: See current code.

File: src/backend/fts/README (hunk 12 above)
Comment (dkrieger-pvtl) [reply src/backend/fts/README]: modified.

File: src/include/cdb/cdbfts.h (hunk 7 above)
Comment (l-wang) [reply src/include/cdb/cdbfts.h]: >Note: unsigned subtraction works properly with wraparound.
>E.g. 255 + 1 = 0, 0 - 255 = 1, 1 - 255 = 2, and so on.

@asimrp as you mentioned, I can see that uint8 overflow wouldn't be a problem for us, however, as @jchampio and @dkrieger-pvtl pointed out, it still doesn't guarantee atomicity and visibility without proper locking. Then using uint8 is just to save a few bytes of memory?

File: src/test/isolation2/input/fts_manual_probe.source [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+--
+-- fts_probe_stats is only queried when the ftsLoop
+-- is stopped at a known location to ensure a consistent view of the stats.
+
+include: helpers/server_helpers.sql;
+
+create extension if not exists gp_inject_fault;
+select gp_inject_fault('all', 'reset', 1);
+
+create temp table fts_probe_results(seq serial, seq_name varchar(20),
```
Comment (l-wang) [inline src/test/isolation2/input/fts_manual_probe.source]: @kalensk @dkrieger-pvtl I'm thinking maybe we can simplify our tests more. These temp tables and views are too specific to our implantation, I can image it's going to be hard to understand for people who don't know the implementation details, and this test is probably going to fail in the future when we replace the busy waiting with condition variable. I just pushed a commit that rewrites the piggyback test without using any UDF, temp table and view. Let me know if it makes sense and if it does we can do something similar for the fresh results case.

File: src/include/cdb/cdbfts.h (hunk 7 above)
Comment (asimrp) [reply src/include/cdb/cdbfts.h]: Yes, my concern indeed boils down to just a few bytes of memory.  But we can flip the question around - if uint8 overflow wouldn't be a problem, why is uint64 needed?

File: src/backend/cdb/cdbfts.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (asimrp) [reply src/backend/cdb/cdbfts.c]: Thank you hearing me out and for the patient explanation.  In times of high load, `FtsNotifyProber()` may become a hot code path and we have not done an assessment of the impact of introducing multiple `SpinlockAcquire()` calls there.  If correctness is not at stake, then performance is what we should care about.

I came across [this commit](https://github.com/greenplum-db/gpdb/commit/a1695035b3612518ee2c5dd2578b769aa6afab93)from upstream today.  Notice the absence of locks and a comment acknowledging it explicitly.
```C
	/*
	 * We don't need to acquire the ckpt_lck in this case because we're only
	 * looking at a single flag bit.
	 */
	if (cps->ckpt_flags & CHECKPOINT_IMMEDIATE)
		return true;
```
  I'm finding it difficult to reason about correctness of such code now, after seeing your and Jacob's comments...

[the remark this answers is not in the document]
Comment (jchampio) [reply src/backend/cdb/cdbfts.c]: > you still need volatile for visibility and ordering for proper synchronization

`volatile` only fixes an order and access method for the compiler; IIUC, it makes no guarantees about the visibility or ordering on the hardware itself. That's why the lock functions exist -- they set up the required memory fences on platforms that need them.

> I came across this commitfrom upstream today. Notice the absence of locks and a comment acknowledging it explicitly.

Maybe they don't need correct ordering/visibility for that particular implementation? I'm skeptical of most "lockfree" solutions; I've seen too many that are plain incorrect in the face of hostile platforms like ARM.

File: src/test/isolation2/input/fts_manual_probe.source (hunk 14 above)
Comment (kalensk) [reply src/test/isolation2/input/fts_manual_probe.source]: The `fts_manual_probe_piggyback` test is essentially the same as our current updated test without checking the start and done counts, and the additional "fresh result" test. 

Currently, the `fts_manual_probe_piggyback` does not guarantee that probe request 1 comes from the same ftsLoop as requests 2 and 3. This is because requests 2 and 3 trigger another ftsLoop to run - the first for loop in ftsNotifyProber which guarantees a new ftsLoop. Thus, it is possible that requests 2 and 3 are only unblocked by this second ftsLoop. We ran into this issue while writing our current test.

To resolve this, we could suspend at `ftsLoop_after_latch` to guarantee that all three requests are unblocked by the first ftsLoop run.

Comment (kalensk) [reply src/test/isolation2/input/fts_manual_probe.source]: The existing test uses the ftsLoop start and done counts to assert the corresponding probes are piggybacking or generating a fresh result. This is a stronger guarantee than the proposed test as it requires reasoning that the results are correct. As mentioned in the [previous comment](https://github.com/greenplum-db/gpdb/pull/8228#discussion_r309832517) while developing the existing test these start and end counts were critical in revealing mistakes in our expectations. For us the counts communicate the intent of the code and test more clearly especially when the test fails.

For these reasons we prefer the existing test.

Comment (l-wang) [reply src/test/isolation2/input/fts_manual_probe.source]: >Currently, the fts_manual_probe_piggyback does not guarantee that probe request 1 comes from the same ftsLoop as requests 2 and 3.

I disagree, the very first gp_request_fts_probe_scan() (**not** request 1 but the one returned before request 1 in the default session) puts the ftsLoop process in waitLatch. The internal interval is already set to an hour, and given this is isolation2 test and no other tests are running at the same time, request 1 is the only request that can trigger the current probe in this test. 

In this test, request 2 and 3 only signals the ftsLoop process after the ftsLoop process is suspended at `ftsLoop_before_probe`, so the current probe is definitely not trigged by request 2 and 3.

Comment (l-wang) [reply src/test/isolation2/input/fts_manual_probe.source]: However, I'm fine with either way for the tests as long as they are stable and serve our purposes today.

Comment (dkrieger-pvtl) [reply src/test/isolation2/input/fts_manual_probe.source]: > I disagree, the very first gp_request_fts_probe_scan() (**not** request 1 but the one returned before request 1 in the default session) puts the ftsLoop process in waitLatch. The internal interval is already set to an hour, and given this is isolation2 test and no other tests are running at the same time, request 1 is the only request that can trigger the current probe in this test.
> 
> In this test, request 2 and 3 only signals the ftsLoop process after the ftsLoop process is suspended at `ftsLoop_before_probe`, so the current probe is definitely not trigged by request 2 and 3.

Yes, you are correct: the ftsLoop is stopped in `ftsLoop_before_wait_latch` while 1, 2 and 3 finish.  So 1, 2 and 3 must all finish due to the ftsLoop triggered by 1.  However, 2 and 3 will trigger another ftsLoop to run immediately after the test is run; they will unset the latch before looping for the current probe to start.  This is fine, just pointing this out.  We mistakenly thought that 2 and 3 could be completed by this extra loop instead of the first one.

We still prefer the existing test over the new one for the reasons given above.

Comment (l-wang) [reply src/test/isolation2/input/fts_manual_probe.source]: Yes, after the test, a new probe will be immediately triggered by 2 and 3, which we don't care.

File: src/backend/cdb/cdbfts.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (l-wang) [reply src/backend/cdb/cdbfts.c]: >In times of high load, FtsNotifyProber() may become a hot code path 

I doubt FtsNotifyProber() is going to become a hot code path in production. Unlike CHECKPOINT, the only external usage of FtsNotifyProber() are tests and our utility tools. I don't think users in production environment would regularly request fts probe themselves. Correctly me if I'm wrong @kalensk @dkrieger-pvtl .

File: src/test/isolation2/input/fts_manual_probe.source (hunk 14 above)
Comment (l-wang) [reply src/test/isolation2/input/fts_manual_probe.source]: I realized that if you remove all the queries that refer the temp table, view, udf, the existing test still makes sense. So might worth mentioning in the test that those are just helper functions and tables that help debugging and explaining what we are testing.

File: src/include/cdb/cdbfts.h (hunk 7 above)
Comment (dkrieger-pvtl) [reply src/include/cdb/cdbfts.h]: We're actually using a `uint32` now.  I suppose we could use a `uint8` but to me it's equivalent and I am used to looking at our code in this way now.  I don't see saving a few bytes as a way to justify spending the effort to make a change like this.  In addition, it's nice to see a larger number in testing, where for a test run, we can get a sense of how long the cluster has been up by seeing a larger tick count.

File: src/test/isolation2/input/fts_manual_probe.source [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+  SELECT seq_name, /* inside a function */
+         start_count AS current_started, /* inside a function */
+         expected_start_delta, /* inside a function */
+         done_count AS current_done, /* inside a function */
+         expected_done_delta /* inside a function */
+  FROM fts_probe_stats(); /* inside a function */
+$$ language sql volatile;
+
+-- ensure the internal regular probes do not affect our test
+!\retcode gpconfig -c gp_fts_probe_interval -v 3600;
```
Comment (asimrp) [inline src/test/isolation2/input/fts_manual_probe.source]: ALTER SYSTEM command is available to accomplish this in a more elegant way in master and 6X, especially when we want to change a master-only GUC.

```SQL
ALTER SYSTEM SET gp_fts_probe_interval TO 3600;
SELECT pg_reload_conf();
```

File: src/test/isolation2/input/fts_manual_probe.source [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+!\retcode gpstop -u;
+
+-- ensure there is no in progress ftsLoop after reloading the gp_fts_probe_interval
+select gp_request_fts_probe_scan();
+select insert_expected_stats('initial', 0, 0);
+select * from get_stats;
+
+-- piggyback test: start multiple probes
+select gp_inject_fault_infinite('ftsNotify_before', 'suspend', 1);
```
Comment (asimrp) [inline src/test/isolation2/input/fts_manual_probe.source]: Please use `gp_inject_fault*2` version of the functions.  That is a newer and better interface that does not use dispatcher to dispatch the fault SQL and `status` fault works reliable in the new interface.

File: src/test/isolation2/input/fts_manual_probe.source (hunk 15 above)
Comment (dkrieger-pvtl) [reply src/test/isolation2/input/fts_manual_probe.source]: `ALTER SYSTEM` does not work properly on our cluster...it just modifies the value on the master AFAIK.  We have `gpconfig` to properly handle the per-instance postgresql.conf files and keep them in sync properly.  I believe the same is true for `pg_reload_conf()`...it just works on master.

So, I'd prefer to keep the `gpconfig; gpstop` idiom, even if we know this guc is only for master.  Although your suggestion is more standard postgresQL, it requires special knowledge for this particular GUC, and might be confusing to others as well.

File: src/test/isolation2/input/fts_manual_probe.source (hunk 16 above)
Comment (dkrieger-pvtl) [reply src/test/isolation2/input/fts_manual_probe.source]: The interface requires `hostname` and `port`, making the resulting queries a bit less elegant to view IMHO.    Since we don't need performance nor `status` fault, is it worth making this change?

Comment (berlin-ab) [reply src/test/isolation2/input/fts_manual_probe.source]: inject_fault*2 interface allows you to ensure that the fault is injected into the acting role for a content id, instead of hardcoding the dbid, which makes the test more resilient.

**Comment by dkrieger-pvtl** [comment]
Based on a comment from @d, we have switched the fts counters from `uint32` to `int32`.   The reasoning is that this will show a wrap-around much easier than a uint32...we'll get obviously a very large negative number.  We do not expect wraparound from an int32 either....

File: src/test/isolation2/input/fts_manual_probe.source (hunk 16 above)
Comment (dkrieger-pvtl) [reply src/test/isolation2/input/fts_manual_probe.source]: But in our case, we expect the cluster to be up without faults. I'm not against the change per se, but just wondering if it's recommended practice to always use `gp_inject_fault2` in favor of `gp_inject_fault`for that reason and because it's considered a more robust implemenation.

File: src/backend/cdb/cdbfts.c [hunk 17]
```diff
@@ -99,21 +99,52 @@ ftsUnlock(void)
...
+			break;
+
+		CHECK_FOR_INTERRUPTS();
 		pg_usleep(50000);
+	}
+
+	/* Wait until current probe in progress is completed */
+	for (;;)
+	{
+		uint32			done;
```
Comment (jchampio) [inline src/backend/cdb/cdbfts.c]: Needs to be `int32`.

(Hopefully there's a new warning from the `>= 0` comparison below?)

Comment (jchampio) [reply src/backend/cdb/cdbfts.c]: Ah, already fixed.

File: src/test/isolation2/input/fts_manual_probe.source [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+-- individual processes. Careful use of fault injectors are needed to have 
+-- complete and consistent control over the flow of the two independent 
+-- processes - the ftsLoop and FtsNotifyProber's.
+--
+-- fts_probe_stats is only queried when the ftsLoop
+-- is stopped at a known location to ensure a consistent view of the stats.
+
+include: helpers/server_helpers.sql;
+
+create extension if not exists gp_inject_fault;
```
[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply src/test/isolation2/input/fts_manual_probe.source]: Maybe we can still leave it in?  Otherwise, during local testing, we have to remember to add it in explicitly, as in:

`./pg_isolation2_regress --load-extension=gp_inject_fault --init-file=init_file_isolation2 fts_manual_probe`

In other words, `gp_inject_fault` is a dependency of this test directly, and that's not captured other tham in `make install`. This seems to cause more trouble than necessary for an operation that will be skipped when run via `make install`(where that module is added).

**Comment by berlin-ab** [comment]
I do prefer making the dependency explicit.

File: src/test/isolation2/input/fts_manual_probe.source (hunk 16 above)
Comment (dkrieger-pvtl) [reply src/test/isolation2/input/fts_manual_probe.source]: Done....switched to `gp_inject_fault2` framework.

File: src/test/isolation2/input/fts_manual_probe.source (hunk 18 above)
[the remark this answers is not in the document]
Comment (pengzhout) [reply src/test/isolation2/input/fts_manual_probe.source]: Hmm, GPDB used to has many "create extension if not exists gp_inject_fault;" in test cases, now they are all removed to use "--load-extension", so my point is just to make the test cases consistent and adding --load-extension=gp_inject_fault to do a manual test is also acceptable to me.

[the remark this answers is not in the document]
Comment (asimrp) [reply src/test/isolation2/input/fts_manual_probe.source]: Hey @dkrieger-pvtl, Pengzhou is correct.  We should remove the create extension statement from the test.  Thank you for catching it Pengzhou.  If we resurrect the practice to create extension like this, it is possible to regress back to an error that was fixed by commit b95b5425706434d2fca7a74ef46c5452ef8418d8.

File: src/include/cdb/cdbfts.h (hunk 7 above)
Comment (dkrieger-pvtl) [reply src/include/cdb/cdbfts.h]: See https://github.com/greenplum-db/gpdb/pull/8228#issuecomment-518293140 . We are using an `int32` now.

File: src/test/isolation2/input/fts_manual_probe.source (hunk 18 above)
[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply src/test/isolation2/input/fts_manual_probe.source]: Done.  Thanks so much for this...I was, in fact, seeing flakes in my test pipelines.

**Comment by dkrieger-pvtl** [comment]
Thanks for all the comments; I squashed all commits together to do one final test before merge as I believe all major comments have been address.  Until this is merged, see https://github.com/dkrieger-pvtl/gpdb/commits/fts_probe_fix_for_review_can_erase for specific commits broken up in a semantic manner(as before), minus the last fix to remove the explicit load of `gp_inject_fault` from the test file.
